### PR TITLE
SCP-4446: Implement buildWithdrawConstraints

### PR DESCRIPTION
**Pull request by bjornkihlberg**

### CODE REVIEW

File: marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs [hunk 1]
```diff
@@ -43,4 +48,9 @@ buildWithdrawConstraints
   :: MarloweVersion v -- ^ The Marlowe version to build the transaction for.
   -> PayoutDatum v -- ^ The role token from which to withdraw funds.
   -> Either WithdrawError (TxConstraints v)
-buildWithdrawConstraints = error "not implemented"
+buildWithdrawConstraints =
+  \case MarloweV1 -> Right . buildWithdrawConstraintsV1
+  where
+    buildWithdrawConstraintsV1 :: AssetId -> TxConstraints 'V1
+    buildWithdrawConstraintsV1 assetId =
+      TxConstraints.mustConsumePayouts assetId <> TxConstraints.mustSpendRoleToken assetId
```
Comment (jhbertra) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:56]: We also need to add a `mustPayToAddress` here to send the funds to the `changeAddress`

Comment (bjornkihlberg) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:56]: Ok, I'll fix it

Comment (bjornkihlberg) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:56]: The `buildWithdrawConstraints` procedure doesn't have access to `Address`. I guess it was incorrectly designed?

Comment (jhbertra) [inline marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:56]: Optional point-free fanciness if you feel like it.

```suggestion
    buildWithdrawConstraintsV1 = TxConstraints.mustConsumePayouts <> TxConstraints.mustSpendRoleToken
```

Comment (jhbertra) [reply marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs:56]: Disregard this comment - transaction balancing will handle this.

File: marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import qualified Language.Marlowe.Runtime.Transaction.Constraints as TxConstraints
+import Test.Hspec (Spec, shouldBe)
+import qualified Test.Hspec as Hspec
+import qualified Test.Hspec.QuickCheck as Hspec.QuickCheck
+import qualified Test.QuickCheck as QuickCheck
+
+base16validStringGen :: QuickCheck.Gen ByteString
+base16validStringGen = do
+  i <- abs <$> QuickCheck.arbitrary
+  fromString <$> QuickCheck.vectorOf (i * 2) (QuickCheck.elements "0123456789abcdef")
```
Comment (jhbertra) [inline marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs:25]:
```suggestion
genBytes :: QuickCheck.Gen ByteString
genBytes = BS.pack <$> arbitrary
```

**Comment by jhbertra** [comment]
You need to run `updateMaterialized` in the repo root.

**Review by jhbertra — APPROVED** [review]
Looks great, thanks!

**Pull request merged**